core/integration/module_test.go

```diff
@@ -2088,222 +2485,437 @@ func (m *Test) Fn() (*Obj, error) {
-			}).
-			WithExec([]string{"sh", "-c", "dagger mod sync"}, dagger.ContainerWithExecOpts{
-				ExperimentalPrivilegedNesting: true,
-			})
+		for _, use := range outer {
+			use := use
 
-		codegenContents, err := modGen.File("/work/dagger.gen.go").Contents(ctx)
-		require.NoError(t, err)
-		require.Contains(t, codegenContents, "Hellov2")
+			t.Run(fmt.Sprintf("%s uses %s", use.sdk, dep.sdk), func(t *testing.T) {
+				t.Parallel()
 
-		newOuter := strings.ReplaceAll(useOuter, `Hello(ctx)`, `Hellov2(ctx)`)
-		modGen = modGen.
-			WithNewFile("/work/main.go", dagger.ContainerWithNewFileOpts{
-				Contents: newOuter,
+				modGen := c.Container().From(golangImage).
+					WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
+					WithWorkdir("/work/dep").
+					With(daggerExec("mod", "init", "--name=dep", "--sdk="+dep.sdk)).
+					With(sdkSource(dep.sdk, dep.source)).
+					WithWorkdir("/work").
+					With(daggerExec("mod", "init", "--name=use", "--sdk="+use.sdk)).
+					With(sdkSource(use.sdk, use.source)).
+					With(daggerExec("mod", "install", "./dep"))
+
+				if use.sdk == "go" {
+					logGen(ctx, t, modGen.Directory("."))
+				}
+
+				out, err := modGen.With(daggerQuery(`{use{useHello}}`)).Stdout(ctx)
+				require.NoError(t, err)
+				require.JSONEq(t, `{"use":{"useHello":"hello"}}`, out)
+
+				// cannot use transitive dependency directly
+				_, err = modGen.With(daggerQuery(`{dep{hello}}`)).Stdout(ctx)
+				require.Error(t, err)
+				require.ErrorContains(t, err, `Cannot query field "dep" on type "Query".`)
 			})
+		}
+	}
+}
 
-		out, err = modGen.With(daggerQuery(`{use{useHello}}`)).Stdout(ctx)
-		require.NoError(t, err)
-		require.JSONEq(t, `{"use":{"useHello":"hello"}}`, out)
-	})
+func TestModuleCodegenOnDepChange(t *testing.T) {
+	t.Parallel()
 
-	t.Run("python uses go", func(t *testing.T) {
-		t.Parallel()
+	c, ctx := connect(t)
 
-		modGen := c.Container().From(golangImage).
-			WithMountedFile(testCLIBinPath, daggerCliFile(t, c)).
-			WithWorkdir("/work/dep").
-			With(daggerExec("mod", "init", "--name=dep", "--sdk=go")).
-			WithNewFile("/work/dep/main.go", dagger.ContainerWithNewFileOpts{
-				Contents: useInner,
-			}).
-			WithWorkdir("/work").
-			With(daggerExec("mod", "init", "--name=use", "--sdk=python")).
-			With(daggerExec("mod", "install", "./dep")).
-			WithNewFile("/work/src/main.py", dagger.ContainerWithNewFileOpts{
-				Contents: `from dagger import dag, function
+	type testCase struct {
+		sdk      string
+		source   string
+		changed  string
+		expected string
+	}
 
-@function
-def use_hello() -> str:
-    return dag.dep().hello()
-`,
-			})
+	for _, tc := range []testCase{
+		{
+			sdk:      "go",
+			source:   useGoOuter,
+			expected: "Hellov2",
+			changed:  strings.ReplaceAll(useGoOuter, `Hello(ctx)`, `Hellov2(ctx)`),
+		},
+		{
+			sdk:      "python",
+			source:   usePythonOuter,
+			expected: "hellov2",
+			changed:  strings.ReplaceAll(usePythonOuter, `.hello()`, `.hellov2()`),
+		},
+	} {
```

Comment:
Similarly to the other comment, first version of this was using permutations of both _go_ and _python_ with _inner_ and _outer_ which is actually a small change to support. But I thought maybe there's not much value in that since what's under test is how a dependency is handled, no matter where it comes from. There's a few other cases like this.

Reply:
I agree with the logic of this - I think maybe we can leave out the matrix of combinations. Especially with typescript tests coming, it would just massively increase the number of tests to run.

Maybe we could take a middle road though if we want to make sure that this code works as we increase the number of SDKs:
- We could try and make sure that every SDK can codegen from a Go API
- We could try and make sure that Go can codegen from every languages Go API

Reply:
I removed the permutation so it's consistent with the other cases now.